include/execution.hpp

```diff
@@ -2171,7 +2173,233 @@ namespace std::execution {
+            __q<__bind_tuples>>;
+
+        using __variant_t =
+          __error_types_of_t<
+            _Sender,
+            make_env_t<get_stop_token_t, stop_token>,
+            __transform<
+              __bind_front_q<__decayed_tuple, set_error_t>,
+              __bound_values_t>>;
+
+        using __receiver = __receiver<__variant_t>;
+
+        stop_source __stop_source_{};
+        example::intrusive_queue<&__operation_base::__next_> __operation_states_;
+        connect_result_t<_Sender, __receiver> __op_state2_;
+        __variant_t __data_;
+
+        mutex __mutex_;
+        __state_t __state_{__state_t::__created};
+
+        explicit __sh_state(_Sender& __sndr)
+          : __op_state2_(
+              connect((_Sender&&) __sndr, __receiver{*this, __stop_source_, __data_}))
+        {
+        }
+
+        void __notify() noexcept override {
+          {
+            lock_guard __lock{__mutex_};
+            __state_ = __state_t::__completed;
+          }
+
+          while(!__operation_states_.empty()) {
+            __operation_states_.pop_front()->__notify();
+          }
+        }
+      };
+
+    template <class _SenderId, class _ReceiverId>
+      class __operation : public __operation_base {
+        using _Sender = __t<_SenderId>;
+        using _Receiver = __t<_ReceiverId>;
+
+        struct __on_stop_requested {
+          stop_source& __stop_source_;
+          void operator()() noexcept {
+            __stop_source_.request_stop();
+          }
+        };
+        using __on_stop = optional<typename stop_token_of_t<
+            env_of_t<_Receiver> &>::template callback_type<__on_stop_requested>>;
+
+        _Receiver __recvr_;
+        __on_stop __on_stop_{};
+        shared_ptr<__sh_state<_Sender>> __shared_state_;
+
+        void __propagate_signal() noexcept {
+          auto &__data = __shared_state_->__data_;
+
+          std::visit([&](auto& __tupl) noexcept -> void {
+            std::apply([&](auto __tag, auto&... __args) noexcept -> void {
+              __tag((_Receiver&&) __recvr_, __args...);
+            }, __tupl);
+          }, __data);
+        }
+
+      public:
+        __operation(_Receiver&& __rcvr,
+                    shared_ptr<__sh_state<_Sender>> __shared_state)
+          : __recvr_((_Receiver&&)__rcvr)
+          , __shared_state_(move(__shared_state)) {
+        }
+
+        void __notify() noexcept override {
+          __on_stop_.reset();
+          __propagate_signal();
+        }
+
+        friend void tag_invoke(start_t, __operation& __self) noexcept try {
+          __sh_state<_Sender> *__shared_state = __self.__shared_state_.get();
+          unique_lock __lock{__shared_state->__mutex_};
+          __state_t __state = __shared_state->__state_;
+
+          if (__state == __state_t::__completed) {
+            __lock.unlock();
+            __self.__propagate_signal();
+          }
+          else {
+            __self.__on_stop_.emplace(
```

Comment:
Registering the callback here while holding a mutex lock is dangerous.
The callback might be executed inline if a stop-request has already been issued.
This could then result in a call to request_stop() which might then go on to call a callback registered by the underlying operation which can then complete synchronously (if this is the second consumer to call start(), the first consumer already calling start() on the underlying operation-state). If it completes synchronously then it will call back to `__sh_state::__notify()` which will then try to lock the mutex again - which is UB.

Having said that, see the other comment for how to implement this without a mutex to totally avoid this issue.

Reply:
The `start()` implementation here would then become something like:
```
void tag_invoke(start_t, __operation& __self) noexcept {
  void* const complete_state = static_cast<void*>(__shared_state.get());
  void* old = __shared_state->__head.load(std::memory_order_acquire);
  if (old != complete_state) {
    __on_stop.emplace(get_stop_token(get_env(__self.__recvr_)), __on_stop_requested{__shared_state->__stop_source_});
  }
  
  do {
    if (old == complete_state) {
      __self.notify();
      return;
    }
    __self.next = static_cast<__operation_base*>(old);
  } while (!__shared_state->__head.compare_exchange_weak(
      old, static_cast<__operation_base*>(__self),
      std::memory_order_release,
      std::memory_order_acquire));
}
```